File: forwarder/app/src/main/java/com/codey/forwarder/CodeyDevicePairingClient.java
package com.codey.app;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.ByteArrayOutputStream;
import java.io.InputStream;
import java.io.OutputStreamWriter;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;

final class CodeyDevicePairingClient {
    private CodeyDevicePairingClient() {}

    static PairingChallenge startPairing(ForwarderSettings settings) throws Exception {
        JSONObject payload = new JSONObject()
            .put("kind", "MOBILE")
            .put("flowType", "codey-mobile-pairing")
            .put("cliName", settings.deviceId)
            .put("requestedBy", settings.deviceId)
            .put("scope", "mobile:pair mobile:whatsapp:ingest mobile:gopay:task");

        HttpResult result = postJson(buildUrl(settings.codeyBaseUrl, "/api/device"), payload, null);
        if (result.statusCode < 200 || result.statusCode > 299) {
            throw new Exception("Pairing start failed: HTTP " + result.statusCode + " " +
                take(result.responseBody, 180));
        }

        JSONObject response = new JSONObject(result.responseBody);
        return new PairingChallenge(
            response.getString("deviceCode"),
            response.getString("userCode"),
            response.optString("verificationUri", "/device"),
            response.optString("verificationUriComplete", ""),
            response.optString("expiresAt", "")
        );
    }

    static PairingResult completePairing(ForwarderSettings settings, String deviceCode)
        throws Exception {
        JSONObject payload = new JSONObject()
            .put("kind", "MOBILE")
            .put("deviceId", settings.deviceId)
            .put("label", settings.deviceId)
            .put("capabilities", new JSONArray()
                .put("whatsapp:ingest")
                .put("gopay:unlink"));

        JSONArray phoneBindings = new JSONArray();
        addPhoneBinding(phoneBindings, settings.whatsappPhoneNumber, "WHATSAPP", "WhatsApp", true);
        addPhoneBinding(phoneBindings, settings.gopayPhoneNumber, "GOPAY", "GoPay", false);
        payload.put("phoneBindings", phoneBindings);

        HttpResult result = postJson(
            buildUrl(settings.codeyBaseUrl, "/api/device/" + deviceCode),
            payload,
            null
        );
        if (result.statusCode < 200 || result.statusCode > 299) {
            throw new Exception("Pairing completion failed: HTTP " + result.statusCode + " " +
                take(result.responseBody, 180));
        }

        JSONObject response = new JSONObject(result.responseBody);
        JSONObject device = response.getJSONObject("device");
        return new PairingResult(
            response.getString("deviceToken"),
            device.getString("id"),
            device.getString("deviceId")
        );
    }

    private static void addPhoneBinding(
        JSONArray bindings,
        String phoneNumber,
        String purpose,
        String label,
        boolean isDefault
    ) throws Exception {
        String normalized = phoneNumber == null ? "" : phoneNumber.trim();
        if (normalized.isEmpty()) {
            return;
        }

        bindings.put(new JSONObject()
            .put("phoneNumber", normalized)
            .put("purpose", purpose)
            .put("label", label)
            .put("isDefault", isDefault));
    }

    private static HttpResult postJson(String url, JSONObject payload, String bearerToken)
        throws Exception {
        HttpURLConnection connection = (HttpURLConnection) new URL(url).openConnection();
        connection.setRequestMethod("POST");
        connection.setConnectTimeout(10_000);
        connection.setReadTimeout(10_000);
        connection.setDoOutput(true);
        connection.setRequestProperty("Content-Type", "application/json; charset=utf-8");
        connection.setRequestProperty("Accept", "application/json");
        if (bearerToken != null && !bearerToken.trim().isEmpty()) {
            connection.setRequestProperty("Authorization", "Bearer " + bearerToken.trim());
        }

        try (OutputStreamWriter writer = new OutputStreamWriter(
            connection.getOutputStream(),
            StandardCharsets.UTF_8
        )) {
            writer.write(payload.toString());
        }

        int statusCode = connection.getResponseCode();
        InputStream stream = statusCode >= 200 && statusCode <= 299
            ? connection.getInputStream()
            : connection.getErrorStream();
        String responseBody = readStream(stream);
        connection.disconnect();
        return new HttpResult(statusCode, responseBody);
    }

    private static String buildUrl(String baseUrl, String path) {
        String normalizedBase = baseUrl == null || baseUrl.trim().isEmpty()
            ? ForwarderConfig.DEFAULT_CODEY_BASE_URL
            : baseUrl.trim();
        while (normalizedBase.endsWith("/")) {
            normalizedBase = normalizedBase.substring(0, normalizedBase.length() - 1);
        }
        return normalizedBase + path;
    }

    private static String readStream(InputStream stream) throws Exception {
        if (stream == null) {
            return "";
        }

        try (InputStream input = stream; ByteArrayOutputStream output = new ByteArrayOutputStream()) {
            byte[] buffer = new byte[4096];
            int read;
            while ((read = input.read(buffer)) != -1) {
                output.write(buffer, 0, read);
            }
            return output.toString(StandardCharsets.UTF_8.name());
        }
    }

    private static String take(String value, int length) {
        if (value == null) {
            return "";
        }
        return value.length() <= length ? value : value.substring(0, length);
    }

    static final class PairingChallenge {
        final String deviceCode;
        final String userCode;
        final String verificationUri;
        final String verificationUriComplete;
        final String expiresAt;

        PairingChallenge(
            String deviceCode,
            String userCode,
            String verificationUri,
            String verificationUriComplete,
            String expiresAt
        ) {
            this.deviceCode = deviceCode;
            this.userCode = userCode;
            this.verificationUri = verificationUri;
            this.verificationUriComplete = verificationUriComplete;
            this.expiresAt = expiresAt;
        }
    }

    static final class PairingResult {
        final String deviceToken;
        final String mobileDeviceId;
        final String deviceId;

        PairingResult(String deviceToken, String mobileDeviceId, String deviceId) {
            this.deviceToken = deviceToken;
            this.mobileDeviceId = mobileDeviceId;
            this.deviceId = deviceId;
        }
    }

    private static final class HttpResult {
        final int statusCode;
        final String responseBody;

        HttpResult(int statusCode, String responseBody) {
            this.statusCode = statusCode;
            this.responseBody = responseBody;
        }
    }
}

File: forwarder/app/src/main/java/com/codey/forwarder/CodeyDevicePairingClient.kt
package com.codey.app

import org.json.JSONArray
import org.json.JSONObject
import java.io.InputStream
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL
import java.nio.charset.StandardCharsets

object CodeyDevicePairingClient {
    fun startPairing(settings: ForwarderSettings): PairingChallenge {
        val payload = JSONObject()
            .put("kind", "MOBILE")
            .put("flowType", "codey-mobile-pairing")
            .put("cliName", settings.deviceId)
            .put("requestedBy", settings.deviceId)
            .put("scope", "mobile:pair mobile:whatsapp:ingest mobile:gopay:task")

        val result = postJson(buildUrl(settings.codeyBaseUrl, "/api/device"), payload)
        if (result.statusCode !in 200..299) {
            throw Exception(
                "Pairing start failed: HTTP ${result.statusCode} ${result.responseBody.takeAtMost(180)}"
            )
        }

        val response = JSONObject(result.responseBody)
        val baseUrl = normalizeBaseUrl(settings.codeyBaseUrl)
        val verificationUri = response.optString("verificationUri", "/device")
        val verificationUriComplete = response.optString("verificationUriComplete", "")
        return PairingChallenge(
            deviceCode = response.getString("deviceCode"),
            userCode = response.getString("userCode"),
            verificationUri = toAbsoluteUrl(baseUrl, verificationUri),
            verificationUriComplete = toAbsoluteUrl(baseUrl, verificationUriComplete),
            expiresAt = response.optString("expiresAt", "")
        )
    }

    fun completePairing(settings: ForwarderSettings, deviceCode: String): PairingResult {
        val payload = JSONObject()
            .put("kind", "MOBILE")
            .put("deviceId", settings.deviceId)
            .put("label", settings.deviceId)
            .put(
                "capabilities",
                JSONArray()
                    .put("whatsapp:ingest")
                    .put("gopay:unlink")
            )

        val phoneBindings = JSONArray()
        addPhoneBinding(phoneBindings, settings.whatsappPhoneNumber, "WHATSAPP", "WhatsApp", true)
        addPhoneBinding(phoneBindings, settings.gopayPhoneNumber, "GOPAY", "GoPay", false)
        payload.put("phoneBindings", phoneBindings)

        val result = postJson(buildUrl(settings.codeyBaseUrl, "/api/device/$deviceCode"), payload)
        if (result.statusCode !in 200..299) {
            throw Exception(
                "Pairing completion failed: HTTP ${result.statusCode} ${result.responseBody.takeAtMost(180)}"
            )
        }

        val response = JSONObject(result.responseBody)
        val device = response.getJSONObject("device")
        return PairingResult(
            deviceToken = response.getString("deviceToken"),
            mobileDeviceId = device.getString("id"),
            deviceId = device.getString("deviceId")
        )
    }

    private fun addPhoneBinding(
        bindings: JSONArray,
        phoneNumber: String?,
        purpose: String,
        label: String,
        isDefault: Boolean
    ) {
        val normalized = phoneNumber?.trim().orEmpty()
        if (normalized.isEmpty()) {
            return
        }

        bindings.put(
            JSONObject()
                .put("phoneNumber", normalized)
                .put("purpose", purpose)
                .put("label", label)
                .put("isDefault", isDefault)
        )
    }

    private fun postJson(
        url: String,
        payload: JSONObject,
        bearerToken: String? = null
    ): HttpResult {
        val connection = URL(url).openConnection() as HttpURLConnection
        connection.requestMethod = "POST"
        connection.connectTimeout = 10_000
        connection.readTimeout = 10_000
        connection.doOutput = true
        connection.setRequestProperty("Content-Type", "application/json; charset=utf-8")
        connection.setRequestProperty("Accept", "application/json")
        if (!bearerToken.isNullOrBlank()) {
            connection.setRequestProperty("Authorization", "Bearer ${bearerToken.trim()}")
        }

        OutputStreamWriter(connection.outputStream, StandardCharsets.UTF_8).use { writer ->
            writer.write(payload.toString())
        }

        val statusCode = connection.responseCode
        val stream = if (statusCode in 200..299) {
            connection.inputStream
        } else {
            connection.errorStream
        }
        val responseBody = readStream(stream)
        connection.disconnect()
        return HttpResult(statusCode, responseBody)
    }

    private fun buildUrl(baseUrl: String?, path: String): String {
        return normalizeBaseUrl(baseUrl) + path
    }

    private fun normalizeBaseUrl(baseUrl: String?): String {
        var normalizedBase = if (baseUrl.isNullOrBlank()) {
            ForwarderConfig.DEFAULT_CODEY_BASE_URL
        } else {
            baseUrl.trim()
        }
        while (normalizedBase.endsWith("/")) {
            normalizedBase = normalizedBase.dropLast(1)
        }
        return normalizedBase
    }

    private fun toAbsoluteUrl(baseUrl: String, value: String): String {
        val trimmed = value.trim()
        if (trimmed.isEmpty()) {
            return ""
        }
        if (
            trimmed.startsWith("http://", ignoreCase = true) ||
            trimmed.startsWith("https://", ignoreCase = true)
        ) {
            return trimmed
        }
        return if (trimmed.startsWith("/")) {
            baseUrl + trimmed
        } else {
            "$baseUrl/$trimmed"
        }
    }

    internal fun normalizeVerificationUrl(baseUrl: String?, value: String): String {
        return toAbsoluteUrl(normalizeBaseUrl(baseUrl), value)
    }

    private fun readStream(stream: InputStream?): String {
        return stream?.use { input ->
            input.readBytes().toString(StandardCharsets.UTF_8)
        }.orEmpty()
    }

    data class PairingChallenge(
        val deviceCode: String,
        val userCode: String,
        val verificationUri: String,
        val verificationUriComplete: String,
        val expiresAt: String
    )

    data class PairingResult(
        val deviceToken: String,
        val mobileDeviceId: String,
        val deviceId: String
    )

    private data class HttpResult(
        val statusCode: Int,
        val responseBody: String
    )
}
